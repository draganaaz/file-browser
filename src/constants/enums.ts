export enum FILE_TYPE {
  FOLDER = 'folder',
  FILE = 'file',
}
