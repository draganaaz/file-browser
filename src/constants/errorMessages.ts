export const ERROR_MESSAGES = {
  INVALID_NAME_OR_EXTENSION: 'Invalid name or extension.',
  ITEM_EXISTS:
    'This item already exists at this location. Please choose a different name.',
  FILE_NOT_FOUND: 'File/Folder not found.',
};
