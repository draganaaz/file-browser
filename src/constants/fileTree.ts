export enum FILE_TYPE {
  FOLDER = 'folder',
  FILE = 'file',
}

export const VALID_EXTENSIONS = {
  TEXT: 'txt',
  JSON: 'json',
  IMAGE: 'png',
};

export const MENU_ITEMS = [
  { label: 'New Folder', type: FILE_TYPE.FOLDER },
  {
    label: 'New Text Document',
    type: FILE_TYPE.FILE,
    extension: VALID_EXTENSIONS.TEXT,
  },
  {
    label: 'New JSON Document',
    type: FILE_TYPE.FILE,
    extension: VALID_EXTENSIONS.JSON,
  },
  {
    label: 'New Image',
    type: FILE_TYPE.FILE,
    extension: VALID_EXTENSIONS.IMAGE,
  },
];
