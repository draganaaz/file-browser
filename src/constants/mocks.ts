import { FileNode } from '../types/FileNode';
import { FILE_TYPE } from './fileTree';

export const initialData: FileNode[] = [
  {
    id: '1',
    name: 'public',
    type: FILE_TYPE.FOLDER,
    children: [
      {
        id: '1-1',
        name: 'exus-logo.png',
        type: FILE_TYPE.FILE,
        fileContent: '/exus.png',
      },
    ],
  },
  {
    id: '2',
    name: 'server',
    type: FILE_TYPE.FOLDER,
    children: [
      {
        id: '2-1',
        name: 'config.json',
        type: FILE_TYPE.FILE,
        fileContent: '{"key": "value"}',
      },
      {
        id: '2-2',
        name: 'notes.txt',
        type: FILE_TYPE.FILE,
        fileContent: 'This is a sample text file.',
      },
    ],
  },
  {
    id: '3',
    name: 'src',
    type: FILE_TYPE.FOLDER,
    children: [],
  },
];
