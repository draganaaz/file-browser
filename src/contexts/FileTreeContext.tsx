import React, {
  createContext,
  Dispatch,
  ReactNode,
  SetStateAction,
  useCallback,
  useContext,
  useState,
} from 'react';
import { FileNode } from '../types/FileNode';
import {
  filterTree,
  findNodeById,
  handleAddNode,
  handleRenameNode,
} from '../utils/treeService';
import { debounce } from '../utils/debounce';
import { FILE_TYPE } from '../constants/fileTree';
import { validateNameAndExtension } from '../utils/validations';

export interface FileTreeContextProps {
  fileTree: FileNode[];
  setFileTree: Dispatch<SetStateAction<FileNode[]>>;
  searchTree: (value: string) => void;
  handleAdd: (parentId: string) => void;
  fileExtension: string | null;
  setFileExtension: Dispatch<SetStateAction<string | null>>;
  itemType: FILE_TYPE | null;
  setItemType: Dispatch<SetStateAction<FILE_TYPE | null>>;
  newItemName: string;
  setNewItemName: Dispatch<SetStateAction<string>>;
  activeParentId: string | null;
  setActiveParentId: Dispatch<SetStateAction<string | null>>;
  closeMenus: () => void;
  errorMessage: string;
  menuNode: string | null;
  setMenuNode: Dispatch<SetStateAction<string | null>>;
  handleRename: (nodeId: string, newName: string) => void;
  isRenaming: boolean;
  setIsRenaming: Dispatch<SetStateAction<boolean>>;
}

const FileTreeContext = createContext<FileTreeContextProps | undefined>(
  undefined
);

export const FileTreeProvider: React.FC<{
  value: FileNode[];
  children: ReactNode;
}> = ({ value: initialData, children }) => {
  const [fileTree, setFileTree] = useState<FileNode[]>(initialData);
  const [filter, setFilter] = useState('');
  const [menuNode, setMenuNode] = useState<string | null>(null);
  const [activeParentId, setActiveParentId] = useState<string | null>(null);
  const [newItemName, setNewItemName] = useState('');
  const [fileExtension, setFileExtension] = useState<string | null>(null);
  const [itemType, setItemType] = useState<FILE_TYPE | null>(null);
  const [errorMessage, setErrorMessage] = useState('');
  const [isRenaming, setIsRenaming] = useState(false);

  const debouncedSearch = useCallback(
    debounce((value: string) => setFilter(value), 500),
    []
  );

  const filteredTree = filterTree(fileTree, filter);

  const closeMenus = useCallback(() => {
    setMenuNode(null);
    setErrorMessage('');
  }, []);

  // Helper function to reset state after operation
  const resetState = () => {
    setActiveParentId(null);
    setNewItemName('');
    setErrorMessage('');
  };

  // Validates and confirms adding a new file or folder, then updates the tree.
  const handleAdd = (parentId: string) => {
    const name = newItemName.trim();
    if (!name) {
      resetState();
      return;
    }

    const validation = validateNameAndExtension(
      name,
      itemType as FILE_TYPE,
      fileExtension
    );
    if (!validation.valid) {
      setErrorMessage(validation.error || 'Invalid name or extension.');
      return;
    }

    const parentFolder = fileTree.find((node) => node.id === parentId);
    if (
      parentFolder?.children?.some(
        (child) => child.name === validation.processedName
      )
    ) {
      setErrorMessage(
        'This item already exists at this location. Please choose a different name.'
      );
      return;
    }

    setFileTree((prevTree) =>
      handleAddNode(
        prevTree,
        parentId,
        validation?.processedName || '',
        itemType as FILE_TYPE
      )
    );
    resetState();
  };

  // Validates and confirms renaming a file or folder, then updates the tree.
  const handleRename = (nodeId: string, newName: string) => {
    setActiveParentId(null);

    const node = findNodeById(fileTree, nodeId);
    if (!node) {
      setErrorMessage('File/Folder not found.');
      return;
    }

    const validation = validateNameAndExtension(
      newName,
      node.type,
      node.name.split('.').pop() || ''
    );
    if (!validation.valid) {
      setErrorMessage(validation.error || 'Invalid name or extension.');
      return;
    }

    setFileTree((prevTree) =>
      handleRenameNode(prevTree, nodeId, validation?.processedName || '')
    );
    setIsRenaming(false);
    setErrorMessage('');
  };

  return (
    <FileTreeContext.Provider
      value={{
        fileTree: filteredTree,
        setFileTree,
        searchTree: debouncedSearch,
        handleAdd,
        fileExtension,
        setFileExtension,
        itemType,
        setItemType,
        newItemName,
        setNewItemName,
        activeParentId,
        setActiveParentId,
        closeMenus,
        errorMessage,
        menuNode,
        setMenuNode,
        handleRename,
        isRenaming,
        setIsRenaming,
      }}
    >
      {children}
    </FileTreeContext.Provider>
  );
};

export const useFileTree = () => {
  const context = useContext(FileTreeContext);

  if (!context) {
    throw new Error('UseFileTree must be used within a FileTreeProvider.');
  }

  return context;
};
