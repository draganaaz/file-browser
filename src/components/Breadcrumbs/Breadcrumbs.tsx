import React, { Dispatch, SetStateAction } from 'react';
import { FileNode } from '../../types/FileNode';
import { useFileTree } from '../../contexts/FileTreeContext';

interface BreadcrumbsProps {
  path: string[];
  setPath: Dispatch<SetStateAction<string[]>>;
}

const Breadcrumbs: React.FC<BreadcrumbsProps> = ({ path, setPath }) => {
  const { fileTree, setSelectedNode } = useFileTree();

  const handleBreadcrumbClick = (index: number) => {
    const newPath: string[] = ['My Files'];
    let currentNode: FileNode | null = null;
    let currentTree = fileTree;

    // Iterate through the breadcrumb segments up to the clicked index
    for (let i = 1; i <= index; i++) {
      const segment = path[i];
      newPath.push(segment);

      currentNode = currentTree.find((node) => node.name === segment) || null;
      currentTree = currentNode?.children || [];
    }

    setPath(newPath);
    setSelectedNode(currentNode);
  };

  return (
    <div
      role="navigation"
      aria-label="Breadcrumbs"
      className="text-gray-500 mb-4"
    >
      {path.map((segment, index) => (
        <span
          key={index}
          className="cursor-pointer"
          onClick={() => handleBreadcrumbClick(index)}
          tabIndex={0}
          role="button"
          onKeyDown={(e) => {
            if (e.key === 'Enter' || e.key === ' ') {
              handleBreadcrumbClick(index);
            }
          }}
          aria-label={`Navigate to ${segment}`}
        >
          {segment} {index < path.length - 1 && ' / '}
        </span>
      ))}
    </div>
  );
};

export default Breadcrumbs;
