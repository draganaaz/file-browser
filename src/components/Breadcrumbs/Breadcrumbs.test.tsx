import React from 'react';
import { render, screen } from '@testing-library/react';
import Breadcrumbs from './Breadcrumbs';
import { useFileTree } from '../../contexts/FileTreeContext';
import { initialData } from '../../constants/mocks';

jest.mock('../../contexts/FileTreeContext', () => ({
  useFileTree: jest.fn(),
}));

const path = ['My Files', 'public', 'images'];
const mockSetPath = jest.fn();
const mockSetSelectedNode = jest.fn();
const mockFileTree = initialData;

describe('Breadcrumbs Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (useFileTree as jest.Mock).mockReturnValue({
      fileTree: mockFileTree,
      setSelectedNode: mockSetSelectedNode,
    });
  });

  it('renders the breadcrumb segments', () => {
    render(<Breadcrumbs path={path} setPath={mockSetPath} />);

    // Use regular expression to match the text segments
    path.forEach((segment) => {
      expect(screen.getByText(new RegExp(segment, 'i'))).toBeInTheDocument();
    });
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <Breadcrumbs path={path} setPath={mockSetPath} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
