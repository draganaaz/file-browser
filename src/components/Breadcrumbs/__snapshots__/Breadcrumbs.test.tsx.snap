// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    aria-label="Breadcrumbs"
    class="text-gray-500 mb-4"
    role="navigation"
  >
    <span
      aria-label="Navigate to My Files"
      class="cursor-pointer"
      role="button"
      tabindex="0"
    >
      My Files  / 
    </span>
    <span
      aria-label="Navigate to public"
      class="cursor-pointer"
      role="button"
      tabindex="0"
    >
      public  / 
    </span>
    <span
      aria-label="Navigate to images"
      class="cursor-pointer"
      role="button"
      tabindex="0"
    >
      images 
    </span>
  </div>
</DocumentFragment>
`;
